src/VisualStudio/Core/Def/Implementation/EditAndContinue/VisualStudioActiveStatementTracker.cs

```diff
@@ -17,9 +18,9 @@ internal sealed class VisualStudioActiveStatementTracker : IManagedActiveStateme
         private readonly IEditAndContinueWorkspaceService _encService;
 
         [ImportingConstructor]
-        public VisualStudioActiveStatementTracker(VisualStudioWorkspace workspace)
+        public VisualStudioActiveStatementTracker(PrimaryWorkspace workspace)
```

Comment:
PrimaryWorkspace is really quite dangerous as there's no guarantee that the Workspace property has yet been initialized if you initialize before the VisualStudioWorkspace. It's use is effectively deprecated. It also creates really icky questions in things like OOP where I might ask "which workspace is primary?"

What was the goal of moving off of VisualStudioWorkspace? Is there a different way you can obtain the applicable workspace?

Reply:
(maybe we need to redo PrimaryWorkspace to be more of an IPrimaryWorkspace and go from there to avoid the screwy initialization. Previously it was a static singleton so was really, really icky.)

Reply:
We said that VisualStudioWorkspace is for local scenarios only. For Nexus we will have another one. I want to remove VisualStudioWorkspace from components that do not depend on it. They just need a workspace service from the workspace that represents the user session. I though that's what the PrimaryWorkspace is. Here, the EnC logic does not have any dependency on VisualStudioWorkspace, it just need a IEditAndContinueWorkspaceService.

I wanted to reuse this code (link it) to debugger test utilities that VS debugger team can use to run EnC tests. Their debug engine test infra does not have dependency on VS. In the tests the workspace would be a EncTestWorkspace that provides the EnC service.

Reply:
Perhaps we should also think about how are we gonna provide access to workspace to our IDE services. If workspace is per user session, in future we might have multiple workspaces in the process since mutliple users might use the same service process. So we'll probably need some kind of WorkspaceProvider, which services can MEF import and ask for a Workspace instance for specific session id that will flow to their APIs. In local case the WorkspaceProvider just returns a singleton VisualStudioWorkspace.